### PR TITLE
Fix browserification & minification

### PR DESCRIPTION
**Pull request by 131**
Browserify fails to resolve the "./http-proxy/" as "./http-proxy/index.js" but as "./http-proxy.js" (so nothing works)
Beeing explicit here does not cost much for http-proxy, yet it's intrinsically complicated for browserify to fix (as trailing slash might be used as a pollyfill shim for native/non-natives addons i.e.  require('url/') vs require('url') )

### CODE REVIEW

**Comment by 131** [comment]
Rebased

File: lib/http-proxy.js [hunk 1]
```diff
@@ -1,7 +1,7 @@
 var http      = require('http'),
     https     = require('https'),
     url       = require('url'),
-    httpProxy = require('./http-proxy/');
+    httpProxy = require('./http-proxy/index.js');
```
Comment (jcrugzz) [inline lib/http-proxy.js]: Could we just add a comment on why this has to happen? Otherwise it may be removed in the future without context

**Comment by 131** [comment]
You got it :cake:

File: lib/http-proxy.js (hunk 1 above)
Comment (twhid) [inline lib/http-proxy.js]: This also breaks other packages depending on [node-resolve](https://github.com/substack/node-resolve/issues/51) (including Jest).

I'd be happy to submit another PR isolated to just this change. Let me know.

**Comment by 131** [comment]
Last two commits are also related to code tooling, when trying to minify the code (through uglify-js), AST fail to detect function scope (& usage) when what's intended here was really simple (just register a list of "named based hooks"). So switching to a plain old vanilla object is working fine and keep the code sane.

**Comment by sorenvind** [comment]
What is the status of this? Sadly this breaks compatibility with Jest :/

**Comment by 131** [comment]
@sorenvind , are you saying my patch breaks some other module ? (or are you waiting for this to be merged ?)

**Comment by jcrugzz** [comment]
Will get this merged in today. Sorry for the wait.

**Comment by 131** [comment]
With the change you requested :+1:

**Comment by jcrugzz** [comment]
Cherry picked. Thanks!

**Comment by twhid** [comment]
:-) thx!

**Comment by sorenvind** [comment]
Wow guys! That was so much faster than I expected! @131 I talked about this being needed to fix the current lack of jest compatibility. 

Thanks!

**Comment by jcrugzz** [comment]
@twhid @sorenvind no problem. I would love to see the real fix in `node-resolve` but it seems that package needs some love.
